src/utils/transaction.ts

```diff
@@ -572,6 +574,124 @@ const transaction = {
+   */
+  async convertTransactionToHistoryTx(
+    tx: Transaction | CreateTokenTransaction | NanoContract,
+    storage: IStorage
+  ): Promise<IHistoryTx> {
+    if (!tx.hash) {
+      throw new Error('To be a history tx a calculated hash is required');
+    }
+
+    const inputs: IHistoryInput[] = [];
+    const outputs: IHistoryOutput[] = [];
+    const txCache: Record<string, IHistoryTx> = {};
+
+    for (const input of tx.inputs) {
+      let spentTx = await storage.getTx(input.hash);
+      if (!spentTx) {
+        // Try cache first
+        if (txCache[input.hash]) {
+          spentTx = txCache[input.hash];
+        } else {
+          // Get from API
+          spentTx = await new Promise((resolve, reject) => {
+            txApi
+              .getTransaction(input.hash, (response: FullNodeTxApiResponse) => {
+                if (!response.success) {
+                  return reject(new Error(response.message ?? ''));
+                }
+
+                if (input.index >= response.tx.outputs.length) {
+                  return reject(new Error('Index outside of tx output array bounds'));
+                }
+                return resolve(this.convertFullNodeTxToHistoryTx(response));
+              })
+              .catch(err => reject(err));
+          });
+
+          if (!spentTx) {
+            // This should not happen since any errors should be treated already.
+            // This if statement is to ensure typing since spentTx starts as IHistoryTx | null.
+            throw new Error('Could not find the spent transaction');
+          }
+          // Update cache
+          txCache[input.hash] = cloneDeep(spentTx);
+        }
+      }
+
+      if (input.index >= spentTx.outputs.length) {
+        throw new Error(
+          `Index (${input.index}) outside of transaction output array bounds (${spentTx.outputs.length})`
+        );
+      }
+
+      const spentOut = spentTx.outputs[input.index];
+      inputs.push({
+        tx_id: input.hash,
+        index: input.index,
+        script: spentOut.script,
+        decoded: spentOut.decoded,
+        token_data: spentOut.token_data,
+        token: spentOut.token,
+        value: spentOut.value,
+      });
+    }
+
+    const tokensArray =
+      tx.version === CREATE_TOKEN_TX_VERSION
+        ? [{ uid: tx.hash }]
+        : tx.tokens.map(tk => ({ uid: tk }));
+    for (const output of tx.outputs) {
+      const script = output.parseScript(storage.config.getNetwork());
+      const out = {
+        value: output.value,
+        token_data: output.tokenData,
+        script: output.script.toString('hex'),
+        decoded: script?.toData() ?? {},
+        spent_by: null, // Cannot reconstruct this field
+      };
+      outputs.push(this.hydrateIOWithToken(out, tokensArray));
+    }
+
+    const histTx: IHistoryTx = {
+      tx_id: tx.hash,
+      signalBits: tx.signalBits,
+      version: tx.version,
+      weight: tx.weight,
+      timestamp: tx.timestamp ?? 0,
```

Comment:
Is it possible not to have a timestamp?

Reply:
It shouldn't, but the types are:
- `Transaction.timestamp: number | null` 
- `IHistoryTx.timestamp: number`

So i'm required to change the "possible" `null` case into a valid number, hence `?? 0`